tests/manage/monitoring/conftest.py

```diff
@@ -0,0 +1,134 @@
+            based on given operation.
+        metadata (dict): This can contain dictionary object with information
+            relevant to test (e.g. volume name, operating host, ...).
+        measure_after (bool): Determine if time measurement is done before or
+            after the operation returns its state. This can be useful e.g.
+            for capacity utilization testing where operation fills capacity
+            and utilized data are measured after the utilization is completed.
+
+    Returns:
+        dict: contains information about `start` and `stop` time of given
+            function and its `result` and provided `metadata`.
+    """
+    def prometheus_log(info, alert_list):
+        """
+        Log all alerts from Prometheus API every 10 seconds.
+
+        Args:
+            run (bool): When this var turns into False the thread stops.
+            alert_list (list): List to be populated with alerts
+        """
+        prometheus = PrometheusAPI()
+        while info.get('run'):
+            alerts_response = prometheus.get(
+                'alerts',
+                payload={
+                    'silenced': False,
+                    'inhibited': False
+                }
+            )
+            assert alerts_response.ok, 'Prometheus API request failed'
+            for alert in alerts_response.json().get('data').get('alerts'):
+                if alert not in alert_list:
+                    logger.info(f"Adding {alert} to alert list")
+                    alert_list.append(alert)
+            time.sleep(10)
+
+    if not measure_after:
+        start_time = time.time()
+
+    # init logging thread that checks for Prometheus alerts
+    # while workload is running
+    # based on https://docs.python.org/3/howto/logging-cookbook.html#logging-from-multiple-threads
+    info = {'run': True}
+    alert_list = []
+
+    logging_thread = threading.Thread(
+        target=prometheus_log,
+        args=(info, alert_list)
+    )
+    logging_thread.start()
+
+    try:
+        result = operation()
+        if measure_after:
+            start_time = time.time()
+        passed_time = time.time() - start_time
+        if minimal_time:
+            additional_time = minimal_time - passed_time
+            if additional_time > 0:
+                time.sleep(additional_time)
+        stop_time = time.time()
+    except KeyboardInterrupt:
+        # Thread should be correctly terminated on next few lines
+        # this is done in case of user interuption to make sure that thread
+        # is terminated correctly
+        pass
+    info['run'] = False
+    logging_thread.join()
+    logger.info(f"Alerts found during measurement: {alert_list}")
+    return {
+        'start': start_time,
+        'stop': stop_time,
+        'result': result,
+        'metadata': metadata,
+        'prometheus_alerts': alert_list
+    }
+
+
+@pytest.fixture(scope="session")
```

Comment:
why session scope?
Isn't this fixture specific for the test case that scale down the mgr deployment replicas?

Reply:
Testing of alerts and metrics is very time consuming. This way it can be used by more tests including metrics. There is also a possibility that in future we will test SNMP, mail or UI alerts. With this design additional tests won't increase testing time exponentially.

Reply:
The idea here is to separate workload and the test itself, so that we can have single workload run, but have multiple test cases evaluating it, possible from different perspectives.